File: app/src/main/java/com/example/mrakopediareader/api/parser/ParserImpl.java
package com.example.mrakopediareader.api.parser;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

abstract class ParserImpl<T> implements Parser<T> {

    @Override
    public ArrayList<T> fromJsonArray(JSONArray array) throws JSONException {
        final int length = array.length();
        final ArrayList<T> results = new ArrayList<>(length);
        for (int i = 0; i < length; ++i) {
            results.add(fromJsonObject(array.getJSONObject(i)));
        }
        return results;
    }
}

File: app/src/main/java/com/example/mrakopediareader/api/parser/ParserImpl.kt
package com.example.mrakopediareader.api.parser

import kotlin.Throws
import org.json.JSONException
import org.json.JSONArray
import java.util.ArrayList

internal abstract class ParserImpl<T> : Parser<T> {
    @Throws(JSONException::class)
    override fun fromJsonArray(jsonArray: JSONArray): ArrayList<T> {
        val length = jsonArray.length()
        val results = ArrayList<T>(length)
        for (i in 0 until length) {
            results.add(fromJsonObject(jsonArray.getJSONObject(i)))
        }
        return results
    }
}